server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java

```diff
@@ -267,7 +267,21 @@ private void addStateHandler(String key, Set<String> keys, LinkedHashSet<String>
             addStateHandler(dependency, keys, ordered, visited);
         }
 
+        for (String dependency : handler.optionalDependencies()) {
+            if (keys.contains(dependency)) {
+                addStateHandler(dependency, keys, ordered, visited);
+            }
+        }
+
         visited.remove(key);
         ordered.add(key);
     }
+
+    /**
```

Comment:
The state handlers are initialized on creation of the service, which is hung off the ActionModule. The Node creation code has many service interdependencies and it's impossible to create all reserved state handlers ahead of time. This API allows us the flexibility to add other state handlers as we build the modules in Node.